design-documents/service-communication-exception-handling.md

```diff
@@ -0,0 +1,107 @@
+## Exception handling in service communication
+
+### Problem
+
+We need to handle network exceptions that may happen in communication between services.
+
+This proposal aims to answer the following questions:
+* Should we use PHP exception class name in API responses or exception codes
+* What should we do with network exceptions (throw or suppress)
+* How do we limit amount of services we need to redeploy when one of the services start to return more exceptions
+
+As we are leaning towards using REST for communication between services, so this proposal will focus on REST.
+
+### Handling exceptions in API best practices
+
+For REST APIs the common practice is to incorporate response code in HTTP header. However, it's not enough for a client to do optimal exception handling as it's not granular enough. This is why most APIs include error message and exception code transferred as a part of message.
+
+#### Examples
+
+Twitter
+
+```
+{
+    "errors": [
+        {
+            "code": 215,
+            "message": "Bad Authentication data."
+        }
+    ]
+}
+```
+
+Facebook
+
+```
+{
+    "error": {
+        "message": "An active access token must be used to query information about the current user.",
+        "type": "OAuthException",
+        "code": 2500,
+        "fbtrace_id": "DzkTMkgIA7V"
+    }
+}
+```
+
+Twilio
+
+```
+<?xml version='1.0' encoding='UTF-8'?>
+<TwilioResponse>
+    <RestException>
+        <Code>20404</Code>
+        <Message>The requested resource /2010-04-01/Accounts/1234/IncomingPhoneNumbers/1234 was not found</Message>
+        <MoreInfo>https://www.twilio.com/docs/errors/20404</MoreInfo>
+        <Status>404</Status>
+    </RestException>
+</TwilioResponse>
+```
+
+### Exception class name or class name
+
+Existing REST API returns only error message, so we need to add exception class name or error code in addition to HTTP response code. This would allow different consumers of the service handle errors or understand what exception need to be handled or which exception type need to be thrown.
+
+Return exception class names
+Pros
+    * No mapping needed, exceptions from API can easily be caught or thrown
+Cons
+    * If consumer or producer service has non PHP or Magento implementation PHP exception class name doesn't make a lot of sense
+    
+Return error code
+Pros
+    * Error codes make sense regardless of technology
+Cons
+    * We would have to maintain some sort of mapping to make it work.
```

Comment:
Now, most of the exceptions will be the same, I think it's not a big deal.

Reply:
I think mapping should be per API endpoint. Anyway, I agree that it's not a big problem.

Reply:
Mapping will still require additional effort for microservice deployment support from our developers and, most importantly, extension developers